### ISSUE TITLE
Native BottomNavigationBar behaviour

### ISSUE THREAD

**Issue by mikebryant71** [lulupointu/vrouter#96]
Hey, great job with the library, I'm just having a few issues implementing a bottom navigation bar. I saw the VNested example which was very helpful however it doesn't really behave like a native UITabBarController.
Here's an example project I did without VRouter, which illustrates the behaviour I'm looking for.
https://github.com/mikebryant71/flutter_navigation

These are the issues I'm having with VRouter:
### Issue 1
**Expected**: When you push a route on the first tab, switch to another tab and then return to the first tab. It should display the pushed route on the first tab.
**Current**: It doesn't restore the pushed route path on the first tab when returning to it. 
```
onTap: (int index) {
   // We can access this username via the local path parameters (stored in VRouteElement)
   final username = VRouter.of(context).pathParameters['username']!;
   if (index == 0) {
     ConnectedRoutes.toProfile(context, username);
   } else {
     ConnectedRoutes.toSettings(context, username);
   }
 }
 ```
          
 **Question:** Do we need to keep track of the current route for each tab manually? And then restore it back here?
 ```
 if (index == 0) { 
      VRouter.of(context).push(currentProfileRoute) // ex: '/profile/detail/update'
 } else  { 
      VRouter.of(context).push(currentSettingsRoute) // ex: '/settings/notifications'
 }
 ```

### Issue 2
**Expected:** When changing tab it doesn't animate and will restore the tab to the state it was in previously without rebuilding.
**Current:** When changing tab it pushes a new route and animates, it also rebuilds that tab.

By disabling the animation you can get it to feel more native, however it still rebuilds the tab when you navigate to it which isn't generally what you'd want for a mobile app. 
**Question:** Is it possible to achieve this kind of behaviour with VRouter, if so, would you be able to point me in the right direction?

Not really a question, more a comment but I noticed a few others were having trouble distinguishing the difference between nested and stacked routes as I was trying to understand it myself. I think I've understood it now. I think it would be helpful though for newcomers to give an example or two of when you would want to use a stacked route vs a nested route and vice versa.

**Comment by lulupointu** [comment]
### Issue 1

This is because push in VRouter is not like push in Navigator 1.
All push does it navigate to a new route.

With this is mind, what you would need is to navigate to the same url you where before for that particular tab. This is something related to state so not built in `VRouter` (VRouter is routing, not state). What you can do however is handle this statefull thing yourself. It's quite easy, here is an example: https://gist.github.com/lulupointu/053954aa66880e26fd9a01ad65a41ef0

The important part of the gist is the `tabsLastPath` in your statefull widget.

### Issue 2 

#### Part 1: The animations

The animations are controlled by the `Page` that you use. When using `VWidget`, you implicitly use `MaterialPage` or `CupertinoPage` (depending on the OS).

However you can easily customize the animations, here is a link on how to do just that: https://vrouter.dev/guide/Examples/Transitions

#### Part 2: The state

As mentioned in **issue 1**, vrouter does not handle the state. This should be done by you using any state management thingy you want.

If what you need is simple enough, you can use `VRouter` with IndexedTabs, though it's not ideal if you have a stack on top of you indexedTabs, would might have to get creative.

<details>
<summary>
Here is the basic concept
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VRouter(
      routes: [
        VWidget(
          key: ValueKey('MainScreen'),
          path: '/',
          widget: MainScreen(index: 0),
        ),
        VWidget(
          key: ValueKey('MainScreen'),
          path: '/settings',
          widget: MainScreen(index: 1),
        ),
      ],
    );
  }
}

class MainScreen extends StatelessWidget {
  final int index;

  const MainScreen({required this.index});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: IndexedStack(
        index: index,
        children: [
          Align(alignment: Alignment.centerLeft, child: MainWidget('HOME')),
          Align(alignment: Alignment.centerRight, child: MainWidget('SETTINGS')),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => context.vRouter.push(index == 0 ? '/settings' : '/'),
      ),
    );
  }
}

class MainWidget extends StatefulWidget {
  final String title;

  const MainWidget(this.title);

  @override
  _MainWidgetState createState() => _MainWidgetState();
}

class _MainWidgetState extends State<MainWidget> {
  var isChecked = false;

  @override
  Widget build(BuildContext context) {
    return IntrinsicWidth(
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(widget.title, style: TextStyle(fontSize: 20)),
            Checkbox(
                value: isChecked, onChanged: (_) => setState(() => isChecked = !isChecked))
          ],
        ),
      ),
    );
  }
}
```
</details>


### Comment on stacked vs nested routes

I know this can be an issue to understand at first this is why this is well documented. But I hear you and I might add:
- A section on vrouter.dev
- An example on vrouter.dev

I don't have much time right now but when I do I will, thanks for the critique

**Comment by mikebryant71** [comment]
Thanks for the quick reply. I've been looking at this today and yeah it seems pretty hard to achieve as it is.

I liked the "Basic Concept" snippet you posted and was able to get it to work nicely with that but when you add a stacked route like you were saying it does pushes the route over the bottom tabs.

I also tried to use a VNester instead which worked better, pushing the routes above the tabs but animation returns once you've pushed a child route on a tab (as it doesn't have the same Key) 
https://streamable.com/myry0i

I tried to get around this by removing the transition manually (not sure if that's the right way):
```
 transitionDuration: Duration.zero,
 buildTransition: (animation, secondaryAnimation, child) => child, 
```

This caused issues internally though when I change tabs
<img width="303" alt="Screen Shot 2021-06-16 at 7 34 32 pm" src="https://user-images.githubusercontent.com/2742756/122195686-ec24ec80-ced9-11eb-8926-948cd162f68e.png">



<details>
<summary>This was the code used in the video</summary>
<p>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

void main() {
  runApp(
    AppContainer(
      child: VRouter(
        debugShowCheckedModeBanner: false,
        routes: [
          VNester(
            path: '/',
            widgetBuilder: (child) =>
                MyScaffold(child), // Child is the widget from nestedRoutes
            nestedRoutes: [
              VWidget(
                  key: ValueKey('MainScreen'),
                  path: null,
                  stackedRoutes: [
                    VWidget(
                        path: 'child',
                        widget: Container(
                            color: Colors.white,
                            child: Center(child: Text('Child widget'))))
                  ],
                  widget: MainScreen(
                    index: 0,
                  )), // null path matches parent
              VWidget(
                key: ValueKey('MainScreen'),
                path: 'settings',
                widget: MainScreen(
                  index: 1,
                ),
                stackedRoutes: [
                  VWidget(
                      path: 'child',
                      widget: Container(
                          color: Colors.red,
                          child: Center(child: Text('test'))))
                ],
              ),
            ],
          ),
        ],
      ),
    ),
  );
}

abstract class BaseWidget extends StatelessWidget {
  String get title;

  String get buttonText;

  String get pushTo;

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(title),
            SizedBox(height: 50),
            ElevatedButton(
              onPressed: () => context.vRouter.push(pushTo),
              child: Text(buttonText),
            ),
            ElevatedButton(
              onPressed: () {
                context.vRouter.push(
                    (pushTo == '/settings') ? '/child' : '/settings/child');
                if (pushTo == '/settings') {
                  AppContainer.lastHomePath = '/child';
                }
              },
              child: Text('Push to child'),
            ),
          ],
        ),
      ),
    );
  }
}

class MyScaffold extends StatelessWidget {
  final Widget child;

  const MyScaffold(this.child);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: child,
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: context.vRouter.url!.contains('settings') ? 1 : 0,
        onTap: (value) => context.vRouter
            .push((value == 0) ? AppContainer.lastHomePath : '/settings'),
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Home'),
          BottomNavigationBarItem(
              icon: Icon(Icons.settings), label: 'Settings'),
        ],
      ),
    );
  }
}

class HomeScreen extends BaseWidget {
  @override
  String get title => 'Home';

  @override
  String get buttonText => 'Go to Settings';

  @override
  String get pushTo => '/settings';
}

class SettingsScreen extends BaseWidget {
  @override
  String get title => 'Settings';

  @override
  String get buttonText => 'Go to Home';

  @override
  String get pushTo => '/';
}

class MainScreen extends StatelessWidget {
  final int index;

  const MainScreen({required this.index});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: IndexedStack(
        index: index,
        children: [HomeScreen(), SettingsScreen()],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => context.vRouter.push(index == 0 ? '/settings' : '/'),
      ),
    );
  }
}

class MainWidget extends StatefulWidget {
  final String title;

  const MainWidget(this.title);

  @override
  _MainWidgetState createState() => _MainWidgetState();
}

class _MainWidgetState extends State<MainWidget> {
  var isChecked = false;

  @override
  Widget build(BuildContext context) {
    return IntrinsicWidth(
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(widget.title, style: TextStyle(fontSize: 20)),
            Checkbox(
                value: isChecked,
                onChanged: (_) => setState(() => isChecked = !isChecked))
          ],
        ),
      ),
    );
  }
}

class AppContainer extends StatelessWidget {
  final Widget child;
  static var lastHomePath = '/';

  AppContainer({Key? key, required this.child}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return child;
  }
}

\```

</p>
</details>

**Comment by lulupointu** [comment]
It seems like a `Duration.zero` transition is too short. But you can decrease it close to zero to achieve the same effect:
```dart
 transitionDuration: Duration.zero,
 buildTransition: (animation, secondaryAnimation, child) => child, 
```

Btw, the **is** the right way to disable every animations.
If you want transition between pages to play/not play, then (you are right again) you should use `key`

**Comment by mikebryant71** [comment]
Yeah unfortunately you can't use the same key for the "stacked routes" so only way is to remove the transition as far as I can tell.
This means you also lose the swipe to pop which is a shame and it would also be quite difficult to only have the animation when you push the route on the tab but not when switching tabs as from VRouter's pov it's the same thing `context.vRouter.push()`.

**Comment by lulupointu** [comment]
I'm happy to say I managed to make something work.

The trick is that what you want is a custom page behavior so you should create your own page.
Basically the page I created is just done by copy pasting everything from flutter `CupertinoPage` but changing the `transitionDuration` to depend on where you come from.
Also note that I override `reverseTransitionDuration`, this is needed because it defaults to `transitionDuration` and this is not what we want, we always want `reverseTransitionDuration` to be the same.

Here is the implementation (extended from the gist I sent you earlier):

```dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

class Book {
  final String title;
  final String author;

  Book(this.title, this.author);
}

main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final List<String> tabsLastPath = [BooksRoutes.defaultPath, ProfileRoutes.defaultPath];

  @override
  Widget build(BuildContext context) {
    return VRouter(
      initialUrl: BooksRoutes.defaultPath,
      routes: [
        VNester(
          path: null,
          widgetBuilder: (child) =>
              ScaffoldScreen(child: child, initialTabsLastPath: tabsLastPath),
          nestedRoutes: [
            VGuard(
              afterEnter: (context, from, to) => tabsLastPath[0] = to,
              afterUpdate: (context, from, to) => tabsLastPath[0] = to,
              stackedRoutes: [BooksRoutes()],
            ),
            VGuard(
              afterEnter: (context, from, to) => tabsLastPath[1] = to,
              afterUpdate: (context, from, to) => tabsLastPath[1] = to,
              stackedRoutes: [ProfileRoutes()],
            ),
          ],
        ),
      ],
    );
  }
}

class BooksRoutes extends VRouteElementBuilder {
  static final _books = '/books';
  static final _book = _books + '/:bookId';

  static void toBooks(BuildContext context) => context.vRouter.push(_books);

  static void toBook(BuildContext context, {required Book book}) =>
      context.vRouter.pushSegments(['books', book.title]);

  static final List<Book> books = [
    Book('Stranger in a Strange Land', 'Robert A. Heinlein'),
    Book('Foundation', 'Isaac Asimov'),
    Book('Fahrenheit 451', 'Ray Bradbury'),
  ];

  static final name = 'books';

  static final defaultPath = _books;

  @override
  List<VRouteElement> buildRoutes() => [
        VWidget(
          path: _books,
          name: name,
          transitionDuration: Duration(milliseconds: 10),
          buildTransition: (_, __, child) => child,
          widget: BooksScreen(books),
          stackedRoutes: [
            VPage(
              pageBuilder: (LocalKey key, Widget child, String? name) =>
                  AnimationAdaptedCupertinoPage(
                name: name,
                child: child,
                key: key,
                transitionDurationBuilder: (BuildContext context) => Duration(
                  milliseconds: context.vRouter.previousUrl == _books ? 300 : 10,
                ),
              ),
              path: _book,
              widget: Builder(
                builder: (context) {
                  final book = books.firstWhere(
                      (book) => book.title == context.vRouter.pathParameters['bookId']);
                  return BookScreen(book);
                },
              ),
            ),
          ],
        ),
      ];
}

class ProfileRoutes extends VRouteElementBuilder {
  static final _profile = '/profile';
  static final _settings = _profile + '/settings';

  static void toProfile(BuildContext context) => context.vRouter.push(_profile);

  static void toSettings(BuildContext context) => context.vRouter.push(_settings);

  static final name = 'profile';

  static final defaultPath = _profile;

  @override
  List<VRouteElement> buildRoutes() => [
        VWidget(
          path: _profile,
          name: name,
          transitionDuration: Duration(milliseconds: 10),
          buildTransition: (_, __, child) => child,
          widget: ProfileScreen(),
          stackedRoutes: [
            VPage(
              pageBuilder: (LocalKey key, Widget child, String? name) =>
                  AnimationAdaptedCupertinoPage(
                name: name,
                child: child,
                key: key,
                transitionDurationBuilder: (BuildContext context) => Duration(
                  milliseconds: context.vRouter.previousUrl == _profile ? 300 : 10,
                ),
              ),
              path: _settings,
              widget: SettingsScreen(),
            ),
          ],
        ),
      ];
}

class ScaffoldScreen extends StatefulWidget {
  final Widget child;
  final List<String> initialTabsLastPath;

  const ScaffoldScreen({
    Key? key,
    required this.child,
    required this.initialTabsLastPath,
  }) : super(key: key);

  @override
  _ScaffoldScreenState createState() =>
      _ScaffoldScreenState(initialTabsLastPath: initialTabsLastPath);
}

class _ScaffoldScreenState extends State<ScaffoldScreen> {
  List<String> tabsLastPath;

  _ScaffoldScreenState({required List<String> initialTabsLastPath})
      : tabsLastPath = initialTabsLastPath;

  @override
  Widget build(BuildContext context) {
    final index = context.vRouter.names.contains(BooksRoutes.name) ? 0 : 1;
    return Scaffold(
      body: widget.child,
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: index,
        onTap: (value) => context.vRouter.push(tabsLastPath[value]),
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.book), label: 'books'),
          BottomNavigationBarItem(icon: Icon(Icons.person), label: 'settings'),
        ],
      ),
    );
  }
}

class BooksScreen extends StatelessWidget {
  final List<Book> books;

  const BooksScreen(this.books);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Books')),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: ListView(
          children: [
            for (var book in books)
              ListTile(
                title: Text(book.title),
                subtitle: Text(book.author),
                onTap: () => BooksRoutes.toBook(context, book: book),
              )
          ],
        ),
      ),
    );
  }
}

class BookScreen extends StatelessWidget {
  final Book book;

  const BookScreen(this.book);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Book')),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              book.title,
              style: Theme.of(context).textTheme.headline6,
            ),
            Text(
              book.author,
              style: Theme.of(context).textTheme.subtitle1,
            ),
          ],
        ),
      ),
    );
  }
}

class ProfileScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Profile')),
      body: Padding(
        padding: const EdgeInsets.all(80.0),
        child: Column(
          mainAxisSize: MainAxisSize.max,
          children: [
            Expanded(
              child: Center(
                child: Text('This is your profile, you can navigate to your settings'),
              ),
            ),
            ElevatedButton(
              onPressed: () => ProfileRoutes.toSettings(context),
              child: Text(
                'Go to settings',
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class SettingsScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Settings')),
      body: Padding(
        padding: const EdgeInsets.all(80.0),
        child: Column(
          children: [
            Expanded(
              child: Center(
                child: Text(
                    'Here are your settings, you can use the button bellow to access a random book'),
              ),
            ),
            ElevatedButton(
              onPressed: () => BooksRoutes.toBook(
                context,
                book: BooksRoutes.books[Random().nextInt(BooksRoutes.books.length)],
              ),
              child: Text(
                'Go to a random book',
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class AnimationAdaptedCupertinoPage extends CupertinoPage {
  final Widget child;
  final String? name;
  final LocalKey key;
  final Duration Function(BuildContext) transitionDurationBuilder;

  AnimationAdaptedCupertinoPage({
    required this.child,
    required this.name,
    required this.key,
    required this.transitionDurationBuilder,
  }) : super(
          child: child,
          name: name,
          key: key,
        );

  @override
  Route createRoute(BuildContext context) {
    return PageBasedCupertinoPageRoute(
      page: this,
      transitionDuration: transitionDurationBuilder(context),
    );
  }
}

class PageBasedCupertinoPageRoute<T> extends PageRoute<T>
    with CupertinoRouteTransitionMixin<T> {
  PageBasedCupertinoPageRoute({
    required CupertinoPage<T> page,
    required this.transitionDuration,
  }) : super(settings: page) {
    assert(opaque);
  }

  @override
  Duration transitionDuration;

  @override
  Duration get reverseTransitionDuration => Duration(milliseconds: 400);

  CupertinoPage<T> get _page => settings as CupertinoPage<T>;

  @override
  Widget buildContent(BuildContext context) => _page.child;

  @override
  String? get title => _page.title;

  @override
  bool get maintainState => _page.maintainState;

  @override
  bool get fullscreenDialog => _page.fullscreenDialog;

  @override
  String get debugLabel => '${super.debugLabel}(${_page.name})';
}
```

Note that the same process could be applied to `MaterialPage` and you could perform a platform check to know which to use.

**Comment by mikebryant71** [comment]
Cool that works pretty nicely. I'm currently looking into what I can do about the other two issues. 
- With VNester using Key trick doesn't work for "caching" the screens when changing tabs.
- Trying to use standard animation when pushing and popping as it looks a bit strange the way it pops due to removing the transition for all cases `buildTransition: (_, __, child) => child,`

**Comment by lulupointu** [comment]
For caching I have not find any solution myself so if you find one please let me know ^^

For the animations in my example I did not remove every single one, just the one we did not need. So I don't understand the issue

**Comment by mikebryant71** [comment]
Will do!

The issue I was describing with the animation is this:
Current: https://streamable.com/xn1wv6
Expected: https://streamable.com/uvduib

Notice the previous screen does not animate when swiping to pop.

**Comment by lulupointu** [comment]
Oh you are right, this is unexpected. I'll have a look into it when I have time.

**Comment by lulupointu** [comment]
I'm sorry but when I use `VPage` and put a cupertinoPage everywhere I can't reproduce the issue. When you are on an emulator you are on IOS right ?

**Comment by lulupointu** [comment]
Oh actually I know why! This is because I use:

```dart
transitionDuration: Duration(milliseconds: 10),
buildTransition: (_, __, child) => child,
```

which effectively uses a custom page instead of a `CupertinoPage`. This is because `CupertinoPage` don't allow us to change the transition duration by default so you have to use a custom page (which is what `buildTransition: (_, __, child) => child,` implies). But of course this means that you loose the animation with Cupertino when popping.

I'll see if I can modify `CupertinoPage` by implementing the route myself in order to pass the transition duration down. I'll keep you updated but do expect this to take a week since I'm quite busy atm.

**Comment by lulupointu** [comment]
Hey, I just "fixed" that in `v1.1.4+16`.

So now you can change the speed of the animation while keeping a `CupertinoRoute`. You just have to use the same code as before, just remove every `buildTransition: (_, __, child) => child,`.

Have a look and let me know how that goes :wink:

**Comment by lulupointu** [comment]
I am closing this for now, feel free to reopen if you notice any issue with the last change I recommended.
